src/main/java/duke/tasklist/TaskList.java

```diff
@@ -0,0 +1,145 @@
+
+    private List<Task> taskList;
+
+    /**
+     * Constructor for the TaskList class.
+     */
+    public TaskList() {
+        this.taskList = new ArrayList<>();
+    }
+
+    /**
+     * Overloaded constructor to allow for loading of previously generated tasks.
+     *
+     * @param taskList Array of existing tasks parsed by Storage
+     */
+    public TaskList(List<Task> taskList) {
+        this.taskList = taskList;
+    }
+
+    /**
+     * Marks a task as complete in the Task List.
+     *
+     * @param index The index of the task to be marked as complete
+     */
+    public String markTaskComplete(int index) {
+        taskList.get(index).markAsDone();
+        String res = "Nice! I've marked this task as having been completed:\n";
+        res += taskList.get(index).toString();
+        return res;
+    }
+
+    /**
+     * Marks a task as incomplete in the Task List.
+     *
+     * @param index The index of the task to be marked as incomplete.
+     */
+    public String markTaskIncomplete(int index) {
+        taskList.get(index).markAsUndone();
+        String res = "Okay, I've marked this task as not done yet:\n";
+        res += taskList.get(index).toString();
+        return res;
+    }
+
+    /**
+     * Adds a task to the Task List.
+     *
+     * @param task The task that is to be added to the List.
+     */
+    public String addTaskToList(Task task) {
+        taskList.add(task);
+        String res = "Got it, i've added this task to your list:\n  " + task + "\n";
+        res += "You now have " + taskList.size() + " tasks in your list.";
+        return res;
+    }
+
+    /**
+     * Deletes a task from the Task List.
+     *
+     * @param index The index of the task to be deleted.
+     */
+    public String deleteTask(int index) {
+        Task removedTask = taskList.get(index);
+        taskList.remove(index);
+        String res = "Okay, I've removed this task:\n";
+        res += removedTask.toString();
+        res += "\nYou now have " + taskList.size() + " tasks in your list.";
+        return res;
+    }
+
+    /**
+     * Searches for tasks in the list that contain any of the search terms.
+     * Search terms are all individual words that come after the [list] keyword.
+     * For an example, "list new test" will return all tasks in the list that have
+     * "new" or "test" in their descriptions.
+     *
+     * The current implementation of search only supports exact matching (case-insensitive).
+     *
+     * @param terms Array of search terms.
+     * @return String that contains all matching tasks, formatted.
+     */
+    public String search(String[] terms) {
+        String result = "";
+        for (Task task: this.taskList) {
+            for (String term: terms) {
+                if (task.toString().toLowerCase().contains(term.toLowerCase())) {
+                   if (!result.contains(task.toString())) {
+                       result += task.toString() + "\n";
+                   }
```

Comment:
Indentations off by one space?
```suggestion
                    if (!result.contains(task.toString())) {
                        result += task.toString() + "\n";
                    }
```